contentful/content-management/changes/20241202-1200-find-and-delete-missing-references.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Contentful migrations can break when entries have links to content that is "Missing or Inaccessible"
+ * This script finds all entries which link to deleted/removed content and removes those references.
+ */
+
+const getClient = require("../helpers/get-client");
+
+module.exports = async function () {
+    const client = getClient();
+    await removeInvalidReferences(client);
+};
+
+/**
+ * Goes through all entries removing invalid links within them
+ *
+ * @param {ClientAPI} client - Contentful management client instance.
+ */
+async function removeInvalidReferences(client) {
+    const contentTypes = await client.contentType.getMany();
+    const entryIds = await getAllValidEntryIds(client);
+    for (const contentType of contentTypes.items) {
+        await removeInvalidReferencesForContentType({
+            client: client,
+            validEntryIds: entryIds,
+            contentType: contentType.sys.id,
+        });
+    }
+}
+
+/**
+ * Finds all valid entryIds within contentful. This has to be split up by content type due to a max limit of 1000 items
+ *
+ * @param {ClientAPI} client - Contentful management client instance.
+ */
+async function getAllValidEntryIds(client) {
+    console.log("Finding all valid entryIds")
+    const contentTypes = await client.contentType.getMany();
+    const entryIds = new Set();
+    for (const contentType of contentTypes.items) {
+        const entries = await client.entry.getMany({
+            query: { limit: 1000, content_type: contentType.sys.id },
+        });
+        entries.items.map((entry) => entryIds.add(entry.sys.id));
+    }
+    return entryIds;
+}
+
+/**
+ * Removes any invalid references for all entries of type contentType in Contentful
+ * This could have been done without querying on contentType, however at most 1000 entries can be queried
+ * at any one time, which isn't enough to return all of them unless limited by type
+ *
+ * @param {ClientAPI} params.client - Contentful management client instance.
+ * @param {Set(string)} params.validEntryIds - ALl valid entry ids.
+ * @param {string} params.contentType - The content type for which entries are being handled.
+ */
+async function removeInvalidReferencesForContentType({
+    client,
+    validEntryIds,
+    contentType,
+}) {
+    console.log(`Processing contentType: ${contentType}`);
+    const entries = await client.entry.getMany({
+        query: { limit: 1000, content_type: contentType },
+    });
+    for (const entry of entries.items) {
```

Comment:
This many nested loops gives me palpations, but I have 0 idea how you could make it better. I mean it even looks good, and is perfectly readable! It's just... loopy.

Reply:
I am also struggling to work out the best option for this - I could break it down into separate methods, processContentType, and then the fields/values but I think it might actually make readability worse in this specific instance, and I don't think something clever with maps is likely to help a lot. I also don't like it, and open to suggestions